### PR TITLE
Merge noggi/synapse/haproxy_avoid_noop_reenable_211

### PR DESCRIPTION
**Pull request by somic**
Merge https://github.com/airbnb/synapse/pull/211

### CODE REVIEW

**Comment by somic** [comment]
Not ready yet, tests are failing.

**Comment by somic** [comment]
@jolynch @jnb @solarkennedy tests pass now; I also fixed a typo in the original PR in the last commit.

File: lib/synapse/config_generator/haproxy.rb [hunk 1]
```diff
@@ -1102,20 +1105,23 @@ def update_backends_at(socket_file_path, watchers)
 
       # parse the stats output to get current backends
       cur_backends = {}
+      re = Regexp.new('^(.+?),(.+?),(?:.*?,){15}(.+?),')
```
Comment (jolynch) [inline lib/synapse/config_generator/haproxy.rb]: This regex scares me a bit ... but ... ok

What's wrong with just splitting on ',' again?

File: lib/synapse/config_generator/haproxy.rb [hunk 2]
```diff
@@ -1136,28 +1142,36 @@ def update_backends_at(socket_file_path, watchers)
...
+      commands = []
+
       # actually enable the enabled backends, and disable the disabled ones
       cur_backends.each do |section, backends|
-        backends.each do |backend|
-          if enabled_backends.fetch(section, []).include? backend
-            command = "enable server #{section}/#{backend}\n"
+        backends.each do |backend, state|
+          if enabled_backends.fetch(section, Set.new).include? backend
+            next if state == 'UP'
```
Comment (jolynch) [inline lib/synapse/config_generator/haproxy.rb]: I think you want starts with UP since backends that are on their way up will be "UP 1/2" e.g.

**Pull request merged**